database.go

```diff
@@ -0,0 +1,723 @@
+package influxdb
+
+import (
+	"encoding/json"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+)
+
+// database is a collection of retention policies and shards. It also has methods
+// for keeping an in memory index of all the measurements, series, and tags in the database.
+// Methods on this struct aren't goroutine safe. They assume that the server is handling
+// any locking to make things safe.
+type database struct {
+	name string
+
+	policies map[string]*RetentionPolicy // retention policies by name
+	shards   map[uint64]*Shard           // shards by id
+
+	defaultRetentionPolicy string
+
+	// in memory indexing structures
+	measurements        map[string]*Measurement // measurement name to object and index
+	seriesToMeasurement map[uint32]*Measurement // map series id to its measurement
+	series              map[uint32]*Series      // map series id to the Series object
+	names               []string                // sorted list of the measurement names
+}
+
+// newDatabase returns an instance of database.
+func newDatabase() *database {
+	return &database{
+		policies:            make(map[string]*RetentionPolicy),
+		shards:              make(map[uint64]*Shard),
+		measurements:        make(map[string]*Measurement),
+		seriesToMeasurement: make(map[uint32]*Measurement),
+		series:              make(map[uint32]*Series),
+		names:               make([]string, 0),
+	}
+}
+
+// shardByTimestamp returns a shard that owns a given timestamp.
+func (db *database) shardByTimestamp(policy string, id uint32, timestamp time.Time) (*Shard, error) {
```

Comment:
Suggestion: it's quite unclear here to what `id` refers, as someone coming across the code for the first time -- after all, we have a lot of IDs in our system. It is worth renaming this variable to `seriesId`?